reference/factorization/factorization_kernels.cpp

```diff
@@ -222,27 +225,103 @@ void initialize_l_u(std::shared_ptr<const ReferenceExecutor> exec,
                 ++current_index_u;
             }
         }
+        // store diagonal values separately
+        auto l_diag_idx = row_ptrs_l[row + 1] - 1;
+        auto u_diag_idx = row_ptrs_u[row];
+        col_idxs_l[l_diag_idx] = row;
+        col_idxs_u[u_diag_idx] = row;
+        vals_l[l_diag_idx] = one<ValueType>();
+        vals_u[u_diag_idx] = diag_val;
     }
 }
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_FACTORIZATION_INITIALIZE_L_U_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void initialize_row_ptrs_l(
+    std::shared_ptr<const ReferenceExecutor> exec,
+    const matrix::Csr<ValueType, IndexType> *system_matrix,
+    IndexType *l_row_ptrs)
+{
+    auto row_ptrs = system_matrix->get_const_row_ptrs();
+    auto col_idxs = system_matrix->get_const_col_idxs();
+    size_type l_nnz{};
+
+    l_row_ptrs[0] = 0;
+    for (size_type row = 0; row < system_matrix->get_size()[0]; ++row) {
+        bool has_diagonal{};
+        for (size_type el = row_ptrs[row]; el < row_ptrs[row + 1]; ++el) {
+            size_type col = col_idxs[el];
+            // skip diagonal
+            l_nnz += col < row;
+        }
+        // add diagonal again
+        l_nnz++;
+        l_row_ptrs[row + 1] = l_nnz;
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_FACTORIZATION_INITIALIZE_ROW_PTRS_L_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void initialize_l(std::shared_ptr<const ReferenceExecutor> exec,
+                  const matrix::Csr<ValueType, IndexType> *system_matrix,
+                  matrix::Csr<ValueType, IndexType> *csr_l, bool diag_sqrt)
+{
+    const auto row_ptrs = system_matrix->get_const_row_ptrs();
+    const auto col_idxs = system_matrix->get_const_col_idxs();
+    const auto vals = system_matrix->get_const_values();
+
+    const auto row_ptrs_l = csr_l->get_const_row_ptrs();
+    auto col_idxs_l = csr_l->get_col_idxs();
+    auto vals_l = csr_l->get_values();
+
+    for (size_type row = 0; row < system_matrix->get_size()[0]; ++row) {
+        size_type current_index_l = row_ptrs_l[row];
+        // if there is no diagonal value, set it to 1 by default
+        auto diag_val = one<ValueType>();
+        for (size_type el = row_ptrs[row]; el < row_ptrs[row + 1]; ++el) {
+            const auto col = col_idxs[el];
+            const auto val = vals[el];
+            if (col < row) {
+                col_idxs_l[current_index_l] = col;
+                vals_l[current_index_l] = val;
+                ++current_index_l;
+            } else if (col == row) {
+                // save diagonal value
+                diag_val = val;
+            }
+        }
+        // store diagonal values separately
+        auto l_diag_idx = row_ptrs_l[row + 1] - 1;
+        col_idxs_l[l_diag_idx] = row;
+        // compute square root with sentinel
+        if (diag_sqrt) {
+            diag_val = sqrt(diag_val);
+            if (!is_finite(diag_val)) {
```

Comment:
The diagonal entry must always be present in the sparsity pattern (so the resulting triangular matrix is non-singular), and if val is close to zero, sqrt(val) should be non-zero too, since then abs(val) < sqrt(abs(val)). Does that make sense? Or did you mean something else?

Reply:
for val close to zero, you are right.

`vals_l[l_diag_idx] = diag_val`
diag_val is set to one if the system_matrix[i, i] is not presented.
If system_matrix[i, i] is presented but equal to 0, the diag_val will be zero here.

Reply:
I believe the difference should not really matter, it only means that we would need one more iteration. But maybe @hartwiganzt can chime in here?

Reply:
If I understand correctly, you are discussing a scenario where the system matrix has a zero on the diagonal - I would argue we allow the algorithm to break in this scenario. So we may not need to consider this situation.

Reply:
I think even if the system matrix has a zero on the diagonal, the ILU does not necessarily have the same property, so this is even less of an issue. A zero on the diagonal could always be updated to a non-zero value during a sweep (in case the exact ILU has a non-zero there)

Reply:
I think it has different result on the same matrix related to representing the zero explicitly.

Reply:
The initial value is different, yes, but after an iteration step, it will most likely be non-zero again. I might be misunderstanding your point though.

Reply:
I see